From 1bf7cadccb0d02627835b9185b2a201665759861 Mon Sep 17 00:00:00 2001
From: Munehisa Kamata <kamatam@amazon.com>
Date: Thu, 24 Aug 2017 22:54:14 +0000
Subject: [PATCH] xen/events: add xen_shutdown_pirqs helper function

Add a simple helper function to "shutdown" active PIRQs, which actually
closes event channels but keeps related IRQ structures intact. PM
suspend/hibernation code will rely on this.

[Srinidhi Rao <srinidhi.rao@broadcom.com> ported this patch to v6.6.x]

Signed-off-by: Munehisa Kamata <kamatam@amazon.com>
Signed-off-by: Anchal Agarwal <anchalag@amazon.com>
Reviewed-by: Munehisa Kamata <kamatam@amazon.com>
Reviewed-by: Eduardo Valentin <eduval@amazon.com>
CR: https://cr.amazon.com/r/8273214/
---
 drivers/xen/events/events_base.c | 12 ++++++++++++
 include/xen/events.h             |  1 +
 2 files changed, 13 insertions(+)

diff --git a/drivers/xen/events/events_base.c b/drivers/xen/events/events_base.c
index 0c5259c68ade..e402e8d12a8c 100644
--- a/drivers/xen/events/events_base.c
+++ b/drivers/xen/events/events_base.c
@@ -2111,6 +2111,18 @@ void xen_irq_resume(void)
 	restore_pirqs();
 }

+void xen_shutdown_pirqs(void)
+{
+	struct irq_info *info;
+
+	list_for_each_entry(info, &xen_irq_list_head, list) {
+		if (info->type != IRQT_PIRQ || !VALID_EVTCHN(info->evtchn))
+			continue;
+
+		shutdown_pirq(irq_get_irq_data(info->irq));
+	}
+}
+
 static struct irq_chip xen_dynamic_chip __read_mostly = {
 	.name			= "xen-dyn",

diff --git a/include/xen/events.h b/include/xen/events.h
index 7488cd51fbf4..43a587b3496c 100644
--- a/include/xen/events.h
+++ b/include/xen/events.h
@@ -85,6 +85,7 @@ static inline void notify_remote_via_evtchn(evtchn_port_t port)
 void notify_remote_via_irq(int irq);

 void xen_irq_resume(void);
+void xen_shutdown_pirqs(void);

 /* Clear an irq's pending state, in preparation for polling on it */
 void xen_clear_irq_pending(int irq);
--
2.35.6

